core/src/main/java/com/linecorp/armeria/internal/common/RequestContextUtil.java

```diff
@@ -48,6 +58,57 @@
     private static final Set<Thread> REPORTED_THREADS =
             Collections.newSetFromMap(new MapMaker().weakKeys().makeMap());
 
+    /**
+     * The default {@link ContextStorage} which stores the {@link RequestContext} in {@link FastThreadLocal}.
+     */
+    public static final ContextStorage defaultContextStorage = new ThreadLocalContextStorage();
+
+    private static final ContextStorage contextStorage;
+
+    static {
+        final List<ContextStorageProvider> providers = ImmutableList.copyOf(
+                ServiceLoader.load(ContextStorageProvider.class));
+        final String contextStorageFqcn = Flags.contextStorage();
+        if (!providers.isEmpty()) {
+            if (providers.size() > 1) {
+                throw new IllegalStateException("Found more than one " +
+                                                ContextStorageProvider.class.getSimpleName() + ". providers:" +
+                                                providers);
+            }
+
+            final ContextStorageProvider provider = providers.get(0);
+            if (!contextStorageFqcn.isEmpty()) {
+                throw new IllegalStateException("Found " + provider + " and " + contextStorageFqcn +
+                                                ". Which one do you want to use?");
```

Comment:
Shouldn't we always prefer what's specified in the flag? How about also logging the actual storage provider type?

Reply:
Let me remove this. 😄